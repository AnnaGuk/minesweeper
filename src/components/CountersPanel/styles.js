import styled from "styled-components";
import { Panel, Counter, Button } from "react95";

export const StyledPanel = styled(Panel)`
  padding: 7px;
  display: flex;
  justify-content: space-between;
`;

export const StyledCounter = styled(Counter)`
  max-width: 50px;
  padding: 4px;
`;

export const SmileyButton = styled.img`
  width: 20px;
  height: 20px;
`;

export const StyledButton = styled(Button)`
  &:after {
    content: none;
  }
`;
