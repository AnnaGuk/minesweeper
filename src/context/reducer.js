import {
  renderBoard,
  getBoardSizeFromBoardVariant,
  revealBoard,
  openEmpty,
  findUnrevealedCells,
  findBugs,
  findFlags,
} from "./helpers";

const createInitState = (variant) => {
  const { bombCount } = getBoardSizeFromBoardVariant(variant);
  const board = renderBoard(variant);
  return {
    gameVariant: variant,
    gameState: "progress",
    minesNumber: bombCount,
    flagsNumber: 0,
    board,
  };
};

export const init = (variant) => {
  return createInitState(variant);
};

export const reducer = (state, action) => {
  switch (action.type) {
    case "openField":
      let status = "progress";
      let updatedBoard = state.board;
      const unrevealedCells = findUnrevealedCells(updatedBoard);
      if (!updatedBoard[action.row][action.col].isFlagged) {
        if (updatedBoard[action.row][action.col].isBomb) {
          status = "lost";
          updatedBoard[action.row][action.col].isColored = true;
          updatedBoard = revealBoard(updatedBoard);
        } else if (updatedBoard[action.row][action.col].isEmpty) {
          openEmpty(action.row, action.col, state.board, state.gameVariant);
        } else updatedBoard[action.row][action.col].isOpen = true;
        if (unrevealedCells.length === state.minesNumber) {
          status = "won";
          updatedBoard = revealBoard(updatedBoard);
        }
      }
      return {
        ...state,
        board: updatedBoard,
        gameState: status,
      };
    case "toggleFlag":
      let flags = state.flagsNumber;
      let bugs = state.minesNumber;
      let gameStatus = state.gameState;
      let newBoard = state.board.map((stateRow) =>
        stateRow.map((stateCol) => {
          if (
            stateCol.row === action.row &&
            stateCol.column === action.col &&
            !stateCol.isOpen
          ) {
            if (stateCol.isFlagged) {
              flags--;
              if (stateCol.isBomb) {
                bugs++;
              }
              return { ...stateCol, isFlagged: false };
            } else {
              flags++;
              if (stateCol.isBomb) {
                bugs--;
              }
              return { ...stateCol, isFlagged: true };
            }
          } else {
            return stateCol;
          }
        })
      );
      if (bugs === 0) {
        let bugArray = findBugs(newBoard);
        let flagArray = findFlags(newBoard);
        if (JSON.stringify(bugArray) === JSON.stringify(flagArray)) {
          gameStatus = "won";
          newBoard = revealBoard(newBoard);
        }
      }
      return {
        ...state,
        board: newBoard,
        flagsNumber: flags,
        minesNumber: bugs,
        gameState: gameStatus,
      };
    case "reset":
      return init(action.variant);
    default:
      return state;
  }
};
