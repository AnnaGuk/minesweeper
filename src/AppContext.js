import React from "react";
import { reducer, init } from "./reducer";

const GameContext = React.createContext([{}, () => {}]);

const GameProvider = (props) => {
  const [state, dispatch] = React.useReducer(reducer, "beginner", init);
  return (
    <GameContext.Provider value={[state, dispatch]}>
      {props.children}
    </GameContext.Provider>
  );
};

export { GameContext, GameProvider };
