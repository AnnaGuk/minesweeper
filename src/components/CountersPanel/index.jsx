import React from "react";
import {
  StyledPanel,
  StyledCounter,
  SmileyButton,
  StyledButton,
} from "./styles";
import smileyFace from "../../assets/smileyface.svg";
import glassesFace from "../../assets/glassesface.svg";
import cryFace from "../../assets/cryface.svg";
import { GameContext } from "../../context/AppContext";

const CountersPanel = () => {
  const [state, dispatch] = React.useContext(GameContext);
  const { minesNumber, gameVariant, flagsNumber } = state;

  return (
    <StyledPanel variant="well">
      <StyledCounter value={minesNumber} size="sm" />
      <StyledButton
        square
        onClick={() => dispatch({ type: "reset", variant: gameVariant })}
      >
        <SmileyButton
          src={
            state.gameState === "won"
              ? glassesFace
              : state.gameState === "lost"
              ? cryFace
              : smileyFace
          }
          alt={
            state.gameState === "won"
              ? ";)"
              : state.gameState === "lost"
              ? ":("
              : ":)"
          }
        />
      </StyledButton>
      <StyledCounter value={flagsNumber} size="sm" />
    </StyledPanel>
  );
};

export default CountersPanel;
