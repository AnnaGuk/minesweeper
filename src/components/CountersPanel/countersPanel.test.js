import { render, cleanup } from "@testing-library/react";
import CountersPanel from "./index.jsx";
import { GameProvider } from "../../context/AppContext";

const renderWithContext = (component) => {
  return {
    ...render(<GameProvider>{component}</GameProvider>),
  };
};

afterEach(cleanup);

test("CountersPanel component snapshot testing", () => {
  const panel = renderWithContext(<CountersPanel />);
  expect(panel).toMatchSnapshot();
});
