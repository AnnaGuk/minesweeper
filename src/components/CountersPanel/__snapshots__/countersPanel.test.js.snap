// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountersPanel component snapshot testing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-bBrOnJ BLhJT sc-fHuLdG cpKmeX"
      >
        <div
          class="sc-dmlrTW cRsllR sc-dtTInj eJxsHo"
        >
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top active"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom active"
            />
            <span
              class="vertical top left active"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left active"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom"
            />
            <span
              class="vertical top left"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top active"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom active"
            />
            <span
              class="vertical top left active"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left active"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
        </div>
        <button
          class="sc-jSgupP fGmjUd sc-XhUPp bmqdNl"
          type="button"
        >
          <img
            alt=":)"
            class="sc-dkIXFM sewxm"
            src="smileyface.svg"
          />
        </button>
        <div
          class="sc-dmlrTW cRsllR sc-dtTInj eJxsHo"
        >
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top active"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom active"
            />
            <span
              class="vertical top left active"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left active"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top active"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom active"
            />
            <span
              class="vertical top left active"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left active"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
          <div
            class="sc-hHftDr gBwTXM"
          >
            <span
              class="horizontal top active"
            />
            <span
              class="center"
            />
            <span
              class="horizontal bottom active"
            />
            <span
              class="vertical top left active"
            />
            <span
              class="vertical top right active"
            />
            <span
              class="vertical bottom left active"
            />
            <span
              class="vertical bottom right active"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-bBrOnJ BLhJT sc-fHuLdG cpKmeX"
    >
      <div
        class="sc-dmlrTW cRsllR sc-dtTInj eJxsHo"
      >
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top active"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom active"
          />
          <span
            class="vertical top left active"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left active"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom"
          />
          <span
            class="vertical top left"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top active"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom active"
          />
          <span
            class="vertical top left active"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left active"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
      </div>
      <button
        class="sc-jSgupP fGmjUd sc-XhUPp bmqdNl"
        type="button"
      >
        <img
          alt=":)"
          class="sc-dkIXFM sewxm"
          src="smileyface.svg"
        />
      </button>
      <div
        class="sc-dmlrTW cRsllR sc-dtTInj eJxsHo"
      >
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top active"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom active"
          />
          <span
            class="vertical top left active"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left active"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top active"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom active"
          />
          <span
            class="vertical top left active"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left active"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
        <div
          class="sc-hHftDr gBwTXM"
        >
          <span
            class="horizontal top active"
          />
          <span
            class="center"
          />
          <span
            class="horizontal bottom active"
          />
          <span
            class="vertical top left active"
          />
          <span
            class="vertical top right active"
          />
          <span
            class="vertical bottom left active"
          />
          <span
            class="vertical bottom right active"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
